frame/elections-phragmen/src/migrations/unlock_and_unreserve_all_funds.rs

```diff
@@ -0,0 +1,494 @@
+			.collect();
+
+		// Total deposited for each account *should* be less than or equal to the total reserved,
+		// however this does not hold for all cases due to bugs in the reserve logic of this pallet.
+		let bugged_deposits = all_accounts
+			.iter()
+			.filter(|account| {
+				account_deposited_sums.get(&account).unwrap_or(&Zero::zero()) >
+					account_reserved_before.get(&account).unwrap_or(&Zero::zero())
+			})
+			.count();
+
+		// The calculation of the locked & staked logic is not known to be bugged, so we assert
+		// that the amounts staked in the pallet match the amounts actually locked in Balances.
+		// If this fails, there is likely a bug in the migration and it should be investigated.
+		ensure!(
+			!all_accounts
+				.iter()
+				.any(|account| account_locked_before.get(&account).unwrap_or(&Zero::zero()) !=
+					account_staked_sums.get(&account).unwrap_or(&Zero::zero())),
+			"account locked != account staked"
+		);
+
+		// Print some summary stats.
+		let total_stake_to_unlock = account_staked_sums.clone().into_values().sum::<BalanceOf<T>>();
+		let total_deposits_to_unreserve =
+			account_deposited_sums.clone().into_values().sum::<BalanceOf<T>>();
+		log::info!("Total accounts: {:?}", all_accounts.len());
+		log::info!("Total stake to unlock: {:?}", total_stake_to_unlock);
+		log::info!("Total deposit to unreserve: {:?}", total_deposits_to_unreserve);
+		log::info!("Bugged deposits: {}/{}", bugged_deposits, all_accounts.len());
+
+		let pre_migration_data =
+			PreMigrationData::<T> { account_locked_before, account_reserved_before };
+		Ok(pre_migration_data.encode())
+	}
+
+	/// Executes the migration.
+	///
+	/// Steps:
+	/// 1. Retrieves the deposit and stake amounts from the pallet.
+	/// 2. Unreserves the deposited funds for each account.
+	/// 3. Unlocks the staked funds for each account.
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		// Get staked and deposited balances as reported by this pallet.
+		let (account_deposited_sums, account_staked_sums, initial_reads_weight) =
+			Self::get_account_deposited_and_staked_sums();
+
+		// Deposited funds need to be unreserved.
+		for (account, unreserve_amount) in account_deposited_sums {
+			if unreserve_amount.is_zero() {
+				continue
+			}
+			T::Currency::unreserve(&account, unreserve_amount);
+		}
+
+		// Staked funds need to be unlocked.
+		for (account, amount) in account_staked_sums {
+			if amount.is_zero() {
+				continue
+			}
+			// Question for reviewers:
+			// remove_lock triggers a few
+			// `ERROR main runtime::system: Logic error: Unexpected underflow in reducing consumer`
+			// errors. are they safe to ignore?
+			//
+			// I have verified that a non-zero lock always exists before remove_lock is called.
+			T::Currency::remove_lock(T::PalletId::get(), &account);
+		}
+
+		// Question for reviewers: how do I know the weight of the unreserve & remove_lock calls?
```

Comment:
> unreserve

There is a [benchmark](https://github.com/paritytech/substrate/blob/master/frame/balances/src/weights.rs#L57) for the `force_unreserve` extrinsic in the balances pallet which is basically does an account lookup and calls `unreserve`. Maybe the corresponding weight can be used here as an aprox?

> remove_lock

We may need to check the implementation in the balances pallet and estimate the expected r/w for removing a lock